lib/common/common.go

```diff
@@ -0,0 +1,350 @@
+	}
+
+	return aciPath, manifest, nil
+}
+
+func ValidateACI(aciPath string) error {
+	aciFile, err := os.Open(aciPath)
+	if err != nil {
+		return err
+	}
+	defer aciFile.Close()
+
+	reader, err := aci.NewCompressedTarReader(aciFile)
+	if err != nil {
+		return err
+	}
+
+	if err := aci.ValidateArchive(reader); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func GenerateManifest(layerData types.DockerImageData, dockerURL *types.ParsedDockerURL) (*schema.ImageManifest, error) {
+	dockerConfig := layerData.Config
+	genManifest := &schema.ImageManifest{}
+
+	appURL := ""
+	// omit docker hub index URL in app name
+	if dockerURL.IndexURL != defaultIndex {
+		appURL = dockerURL.IndexURL + "/"
+	}
+	appURL += dockerURL.ImageName + "-" + layerData.ID
+	appURL, err := appctypes.SanitizeACName(appURL)
+	if err != nil {
+		return nil, err
+	}
+	name, err := appctypes.NewACName(appURL)
+	if err != nil {
+		return nil, err
+	}
+	genManifest.Name = *name
+
+	acVersion, _ := appctypes.NewSemVer(schemaVersion)
+	genManifest.ACVersion = *acVersion
+
+	genManifest.ACKind = appctypes.ACKind("ImageManifest")
+
+	var labels appctypes.Labels
+	var parentLabels appctypes.Labels
+
+	layer, _ := appctypes.NewACName("layer")
+	labels = append(labels, appctypes.Label{Name: *layer, Value: layerData.ID})
+
+	tag := dockerURL.Tag
+	version, _ := appctypes.NewACName("version")
+	labels = append(labels, appctypes.Label{Name: *version, Value: tag})
+
+	if layerData.OS != "" {
+		os, _ := appctypes.NewACName("os")
+		labels = append(labels, appctypes.Label{Name: *os, Value: layerData.OS})
+		parentLabels = append(parentLabels, appctypes.Label{Name: *os, Value: layerData.OS})
+
+		if layerData.Architecture != "" {
+			arch, _ := appctypes.NewACName("arch")
+			parentLabels = append(parentLabels, appctypes.Label{Name: *arch, Value: layerData.Architecture})
+		}
+	}
+
+	genManifest.Labels = labels
+
+	if dockerConfig != nil {
+		exec := getExecCommand(dockerConfig.Entrypoint, dockerConfig.Cmd)
+		if exec != nil {
+			user, group := parseDockerUser(dockerConfig.User)
+			var env appctypes.Environment
+			for _, v := range dockerConfig.Env {
+				parts := strings.SplitN(v, "=", 2)
+				env.Set(parts[0], parts[1])
+			}
+			app := &appctypes.App{
+				Exec:             exec,
+				User:             user,
+				Group:            group,
+				Environment:      env,
+				WorkingDirectory: dockerConfig.WorkingDir,
+			}
+			genManifest.App = app
+		}
+	}
+
+	if layerData.Parent != "" {
+		var dependencies appctypes.Dependencies
+		parentAppNameString := dockerURL.IndexURL + "/" + dockerURL.ImageName + "-" + layerData.Parent
+		parentAppNameString, err := appctypes.SanitizeACName(parentAppNameString)
+		if err != nil {
+			return nil, err
+		}
+		parentAppName, err := appctypes.NewACName(parentAppNameString)
+		if err != nil {
+			return nil, err
+		}
+
+		dependencies = append(dependencies, appctypes.Dependency{App: *parentAppName, Labels: parentLabels})
```

Comment:
something is weird here, maybe got lost in a historical refactoring, but afaict `var dependencies` can be scrapped and you can just assign directly to `genManifest.Dependencies = appctypes.Dependencies{appctypes.Dependency{App: *parentAppName, Labels: parentLabels})`